### PR TITLE
fix: team provider migration

### PR DESCRIPTION
**Pull request by ammarkarachi**
*Issue #, if available:*

*Description of changes:*

The goal of the PR is to change how we mange third party secrets. The secrets are stored in an ephemeral file called ~/,aws/amplify/deployment-secrets.json. On push-resource the auth for which the secrets are deployed. The change in index.ts intercepts incoming commands to find out if the user has to go through migration.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-cli/src/utils/team-provider-migrate.ts
Comment (ammarkarachi) [inline packages/amplify-cli/src/utils/team-provider-migrate.ts:5]: @renebrandel I will need your input for the migration message

File: packages/amplify-category-auth/resources/cloudformation-templates/auth-template.yml.ejs
Comment (kaustavghosh06) [inline packages/amplify-category-auth/resources/cloudformation-templates/auth-template.yml.ejs:324]: I don't think this is your change. Can you rebase?

File: packages/amplify-category-auth/resources/cloudformation-templates/auth-template.yml.ejs
Comment (kaustavghosh06) [inline packages/amplify-category-auth/resources/cloudformation-templates/auth-template.yml.ejs:46]: You'd need to set the default value to NONE in the template itself I think or are you dynamically passing this value during deploy time as an input param?

File: packages/amplify-category-auth/resources/cloudformation-templates/auth-template.yml.ejs
Comment (kaustavghosh06) [inline packages/amplify-category-auth/resources/cloudformation-templates/auth-template.yml.ejs:569]: What if customers have the "providerMeta" values set - are those values ignored here? The prviderMeta is configured below in the same lambda custom resource.

File: packages/amplify-cli-core/src/state-manager/stateManager.ts
Comment (kaustavghosh06) [inline packages/amplify-cli-core/src/state-manager/stateManager.ts:33]: can we store this permission level as a constant. I think we re-use this level of permissions during `amplify configure` as well

File: packages/amplify-provider-awscloudformation/src/push-resources.js
Comment (kaustavghosh06) [inline packages/amplify-provider-awscloudformation/src/push-resources.js:160]: Make sure you filter of the provider - "awscloudformation" and the service "Cognito" as well here.

File: packages/amplify-cli-core/src/state-manager/stateManager.ts
Comment (kaustavghosh06) [inline packages/amplify-cli-core/src/state-manager/stateManager.ts:76]: FromTeamProviderToDeploymentSecrets right?

**Review by kaustavghosh06 — COMMENTED** [review]
In your description could you please describe the changes a bit more? Like whats the file name and what's the structure of this file?

File: packages/amplify-category-auth/resources/cloudformation-templates/auth-template.yml.ejs
Comment (ammarkarachi) [reply packages/amplify-category-auth/resources/cloudformation-templates/auth-template.yml.ejs:324]: No it's not

File: packages/amplify-category-auth/resources/cloudformation-templates/auth-template.yml.ejs
Comment (ammarkarachi) [reply packages/amplify-category-auth/resources/cloudformation-templates/auth-template.yml.ejs:46]: I am dynamically deploying this value if it exists

File: packages/amplify-category-auth/resources/cloudformation-templates/auth-template.yml.ejs
Comment (ammarkarachi) [reply packages/amplify-category-auth/resources/cloudformation-templates/auth-template.yml.ejs:569]: Do we want to update the provider meta if there are no secrets present?

File: packages/amplify-category-auth/resources/cloudformation-templates/auth-template.yml.ejs
Comment (ammarkarachi) [reply packages/amplify-category-auth/resources/cloudformation-templates/auth-template.yml.ejs:46]: I am not entirely sure if `NONE` can be added as a default value 
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html

File: packages/amplify-category-auth/resources/cloudformation-templates/auth-template.yml.ejs
Comment (ammarkarachi) [reply packages/amplify-category-auth/resources/cloudformation-templates/auth-template.yml.ejs:324]: I rebase there seems to be a bad merge I can checkout re commit

**Review by attilah — CHANGES_REQUESTED** [review]
Overall the approach looks ok, but I have a concern about connecting the secrets stuff to the AmplifyAppId which in case of 25+ Apps will not be present and I think the current  implementation would fail. since we don't have a project id or something like that, what if we would use the root stack's resource Id (last part of the ARN after the /), that is guaranteed to be unique by CF, right?

Based on this the suggested new JSON structure would have the `rootStackId` instead of `amplifyAppId`, and we could lose the environment name at all if we want since rootStackId would give a different granularity to the secrets.

A requested also changes around `StateManager` and tests. 

Proposed structure with root stack id: 
```json
{
    "appSecrets": [
        {
            "rootStackId": "<uuid>",
            // However it will be always a single env this way, but keep for portable future changes
            "environments": {
                "dev": {
                    "auth": {
                        "teamprovider1819bdce": {
                            "hostedUIProviderCreds": "[{'ProviderName':'Facebook','client_id':'asdasdasdasd','client_secret':'asdasdasd'}]",
                        }
                    }
                }
            }
        }
    ]
}
```

File: packages/amplify-category-auth/resources/cloudformation-templates/auth-template.yml.ejs
Comment (ammarkarachi) [reply packages/amplify-category-auth/resources/cloudformation-templates/auth-template.yml.ejs:570]: After a deployment we discard the secret and send out an empty array

File: packages/amplify-cli-core/src/state-manager/stateManager.ts
Comment (ammarkarachi) [reply packages/amplify-cli-core/src/state-manager/stateManager.ts:32]: Can you elaborate on making the methods the same? I think it's the same like here
``` json
  setProjectConfig = (projectPath: string | undefined, projectConfig: $TSAny): void => {
    const filePath = pathManager.getProjectConfigFilePath(projectPath);

    JSONUtilities.writeJson(filePath, projectConfig);
  };
```

File: packages/amplify-console-integration-tests/__tests__/pullAndInit.test.ts
Comment (ammarkarachi) [reply packages/amplify-console-integration-tests/__tests__/pullAndInit.test.ts:204]: I do have a test that does this should I write a new test?

File: packages/amplify-cli/src/utils/team-provider-migrate.ts
Comment (ammarkarachi) [reply packages/amplify-cli/src/utils/team-provider-migrate.ts:12]: It's updating the auth if it exists with moving the team provider info and updating the CFN with already existing values

File: packages/amplify-cli/src/utils/team-provider-migrate.ts
Comment (edwardfoyle) [inline packages/amplify-cli/src/utils/team-provider-migrate.ts:21]:
```suggestion
  return context?.exeInfo?.inputParams?.yes
```

I would also rename this to `isYesFlagSet` and move this function to a generic utils file

File: packages/amplify-cli/src/utils/team-provider-migrate.ts
Comment (edwardfoyle) [inline packages/amplify-cli/src/utils/team-provider-migrate.ts:7]: function name should start with lowercase. Also use arrow notation. Also would be nice to have a short comment on what the return value means (looks like it just returns whether or not the migration logic was executed)

File: packages/amplify-cli-core/src/state-manager/stateManager.ts
Comment (edwardfoyle) [inline packages/amplify-cli-core/src/state-manager/stateManager.ts:63]: return type should be boolean

File: packages/amplify-cli-core/src/state-manager/stateManager.ts
Comment (edwardfoyle) [inline packages/amplify-cli-core/src/state-manager/stateManager.ts:70]:
```suggestion
          return !!envTeamProvider.categories.auth[resource][hostedUIProviderCredsField];
```

File: packages/amplify-cli-core/src/state-manager/stateManager.ts
Comment (edwardfoyle) [inline packages/amplify-cli-core/src/state-manager/stateManager.ts:86]: same here, filter before the forEach. Also, if only one match is expected (which I think is the case) you can use `.find()` instead of `.filter()`. Can also use find above as there should only be one auth key as well

File: packages/amplify-cli-core/src/state-manager/stateManager.ts
Comment (edwardfoyle) [inline packages/amplify-cli-core/src/state-manager/stateManager.ts:82]: const

File: packages/amplify-cli-core/src/state-manager/stateManager.ts
Comment (edwardfoyle) [inline packages/amplify-cli-core/src/state-manager/stateManager.ts:89]: don't need to assign back to secrets

File: packages/amplify-cli/src/extensions/amplify-helpers/envResourceParams.ts
Comment (edwardfoyle) [inline packages/amplify-cli/src/extensions/amplify-helpers/envResourceParams.ts:87]: Would there ever be a case where a value in team provider would overwrite a value from deployment secrets when we didn't want to overwrite?

File: packages/amplify-cli-core/src/errors/index.ts
Comment (edwardfoyle) [inline packages/amplify-cli-core/src/errors/index.ts:13]: TeamProviderInfoMigraionError

File: packages/amplify-cli/src/utils/team-provider-migrate.ts
Comment (edwardfoyle) [reply packages/amplify-cli/src/utils/team-provider-migrate.ts:5]: Does the customer really have a choice? What happens if they say no? Should this decision be behind a FF instead of a prompt?

Rene should have the final say in the experience here, but as a customer I think I would be confused about how to answer this question

**Comment by ammarkarachi** [comment]
> Overall the approach looks ok, but I have a concern about connecting the secrets stuff to the AmplifyAppId which in case of 25+ Apps will not be present and I think the current implementation would fail. since we don't have a project id or something like that, what if we would use the root stack's resource Id (last part of the ARN after the /), that is guaranteed to be unique by CF, right?
> 
> Based on this the suggested new JSON structure would have the `rootStackId` instead of `amplifyAppId`, and we could lose the environment name at all if we want since rootStackId would give a different granularity to the secrets.
> 
> A requested also changes around `StateManager` and tests.
> 
> Proposed structure with root stack id:
> 
> ```json
> {
>     "appSecrets": [
>         {
>             "rootStackId": "<uuid>",
>             // However it will be always a single env this way, but keep for portable future changes
>             "environments": {
>                 "dev": {
>                     "auth": {
>                         "teamprovider1819bdce": {
>                             "hostedUIProviderCreds": "[{'ProviderName':'Facebook','client_id':'asdasdasdasd','client_secret':'asdasdasd'}]",
>                         }
>                     }
>                 }
>             }
>         }
>     ]
> }
> ```
I think the 25 AppId limit is an acceptable limit risk. It would be difficult to change the design after it has been approved and implemented

File: package.json
Comment (ammarkarachi) [reply package.json:32]: Yeah I will drop the commit

File: .circleci/config.base.yml
Comment (ammarkarachi) [reply .circleci/config.base.yml:784]: Yeah I will drop the commit

File: packages/amplify-cli/src/extensions/amplify-helpers/envResourceParams.ts
Comment (ammarkarachi) [reply packages/amplify-cli/src/extensions/amplify-helpers/envResourceParams.ts:87]: No not for new projects and migrated projects

**Comment by edwardfoyle** [comment]
> > Overall the approach looks ok, but I have a concern about connecting the secrets stuff to the AmplifyAppId which in case of 25+ Apps will not be present and I think the current implementation would fail. since we don't have a project id or something like that, what if we would use the root stack's resource Id (last part of the ARN after the /), that is guaranteed to be unique by CF, right?
> > Based on this the suggested new JSON structure would have the `rootStackId` instead of `amplifyAppId`, and we could lose the environment name at all if we want since rootStackId would give a different granularity to the secrets.
> > A requested also changes around `StateManager` and tests.
> > Proposed structure with root stack id:
> > ```json
> > {
> >     "appSecrets": [
> >         {
> >             "rootStackId": "<uuid>",
> >             // However it will be always a single env this way, but keep for portable future changes
> >             "environments": {
> >                 "dev": {
> >                     "auth": {
> >                         "teamprovider1819bdce": {
> >                             "hostedUIProviderCreds": "[{'ProviderName':'Facebook','client_id':'asdasdasdasd','client_secret':'asdasdasd'}]",
> >                         }
> >                     }
> >                 }
> >             }
> >         }
> >     ]
> > }
> > ```
> 
> I think the 25 AppId limit is an acceptable limit risk. It would be difficult to change the design after it has been approved and implemented

Even if we stick with using appId as the identifier, I would still like the json to be structured as Attila mentioned. To me having some random string of characters be a key in the json just looks messy and it doesn't convey any meaning

File: packages/amplify-cli/src/__tests__/team-provider-migrate.test.ts
Comment (ammarkarachi) [reply packages/amplify-cli/src/__tests__/team-provider-migrate.test.ts:87]: These are 

> > > Overall the approach looks ok, but I have a concern about connecting the secrets stuff to the AmplifyAppId which in case of 25+ Apps will not be present and I think the current implementation would fail. since we don't have a project id or something like that, what if we would use the root stack's resource Id (last part of the ARN after the /), that is guaranteed to be unique by CF, right?
> > > Based on this the suggested new JSON structure would have the `rootStackId` instead of `amplifyAppId`, and we could lose the environment name at all if we want since rootStackId would give a different granularity to the secrets.
> > > A requested also changes around `StateManager` and tests.
> > > Proposed structure with root stack id:
> > > ```json
> > > {
> > >     "appSecrets": [
> > >         {
> > >             "rootStackId": "<uuid>",
> > >             // However it will be always a single env this way, but keep for portable future changes
> > >             "environments": {
> > >                 "dev": {
> > >                     "auth": {
> > >                         "teamprovider1819bdce": {
> > >                             "hostedUIProviderCreds": "[{'ProviderName':'Facebook','client_id':'asdasdasdasd','client_secret':'asdasdasd'}]",
> > >                         }
> > >                     }
> > >                 }
> > >             }
> > >         }
> > >     ]
> > > }
> > > ```
> > 
> > 
> > I think the 25 AppId limit is an acceptable limit risk. It would be difficult to change the design after it has been approved and implemented
> 
> Even if we stick with using appId as the identifier, I would still like the json to be structured as Attila mentioned. To me having some random string of characters be a key in the json just looks messy and it doesn't convey any meaning

I will try but this will be difficult seeing that deadlines is really close and this requires a change through most of the implementation. This would have been the two design meeting we had regarding this.

**Comment by kaustavghosh06** [comment]
@edwardfoyle as a first step - this is moving all the existing structure which is in team-provider to this file. We can do an API review at a later stage if we want to change the structure. Right now in the migration step we're copy pasting what we have in team-provider to out here and using it only for deployments.

File: packages/amplify-cli/src/utils/team-provider-migrate.ts
Comment (edwardfoyle) [inline packages/amplify-cli/src/utils/team-provider-migrate.ts:17]: it looks like it should be fine because this condition checks for `!isPulling()` but can you double check that this headless pull command still works with your changes: https://github.com/aws-amplify/amplify-cli/issues/3642#issuecomment-655804757

File: .circleci/config.yml
Comment (edwardfoyle) [inline .circleci/config.yml:405]: I don't think this env var should be set. It is only intended for testing the packaged CLI. The tests should be able to find the node CLI without it

File: packages/amplify-cli-core/src/deploymentSecrestHelper.ts
Comment (edwardfoyle) [inline packages/amplify-cli-core/src/deploymentSecrestHelper.ts:1]: typo in filename "secrets"

**Review by nikhname — COMMENTED** [review]
This PR doesn't seem to contain the changes to getCLIPath we discussed. Are we planning on having those in a separate PR?

File: packages/amplify-cli-core/src/index.ts
Comment (edwardfoyle) [inline packages/amplify-cli-core/src/index.ts:59]: no $ on type. The $ denotes it's an incomplete type (has anys in it)

**Comment by ammarkarachi** [comment]
> This PR doesn't seem to contain the changes to getCLIPath we discussed. Are we planning on having those in a separate PR?
@nikhname 
https://github.com/aws-amplify/amplify-cli/blob/6509e6b91774633dc9cfbf4508bba58e11b0bf43/.circleci/config.base.yml#L355
we use AMPLIFY_PATH for that

File: packages/amplify-cli/src/utils/team-provider-migrate.ts
Comment (ammarkarachi) [reply packages/amplify-cli/src/utils/team-provider-migrate.ts:17]: I did a headless pull and init

File: .circleci/config.yml
Comment (ammarkarachi) [reply .circleci/config.yml:405]: It does not pick up the right binary and the migration tests right now are broken if you actually point it 4.0.0

Comment (edwardfoyle) [reply .circleci/config.yml:405]: ah yeah I see you're using it to migration test. that's fine then

File: packages/amplify-cli/src/utils/team-provider-migrate.ts
Comment (edwardfoyle) [reply packages/amplify-cli/src/utils/team-provider-migrate.ts:17]: a headless pull with the social secrets specified in --categories?

Comment (ammarkarachi) [reply packages/amplify-cli/src/utils/team-provider-migrate.ts:19]: Yes

File: packages/amplify-cli/src/utils/team-provider-migrate.ts
Comment (edwardfoyle) [reply packages/amplify-cli/src/utils/team-provider-migrate.ts:28]: saw your comment in the latest commit. makes sense now

**Review by edwardfoyle — APPROVED** [review]

File: packages/amplify-cli/src/utils/team-provider-migrate.ts
Comment (ammarkarachi) [reply packages/amplify-cli/src/utils/team-provider-migrate.ts:28]: // return true if the current state of the app does not contain secrets in the team provider info
// if the state of the app is not without secrets in team-provider-info it return false

File: packages/amplify-cli/src/utils/team-provider-migrate.ts
Comment (ammarkarachi) [reply packages/amplify-cli/src/utils/team-provider-migrate.ts:19]: there is a tagged release if you would like to make sure `@teamprv2`

**Review by nikhname — APPROVED** [review]

**Review by attilah — APPROVED** [review]
LGTM, thanks for the changes, just update `enableExternalAuth` as we discussed, and store the root stack id, beside `amplifyAppId`, in case we need it in the future.

**Pull request merged**
